### PR TITLE
On decorator

### PR DESCRIPTION
**Pull request by willmcgugan**
Implement the @on decorator

Fixes https://github.com/Textualize/textual/issues/2398

### CODE REVIEW

File: docs/guide/events.md [hunk 1]
```diff
@@ -155,6 +155,66 @@ Textual uses the following scheme to map messages classes on to a Python method.
 --8<-- "docs/images/events/naming.excalidraw.svg"
 </div>
 
+### On decorator
+
+In addition to the naming convention, message handlers may be created with the [on][textual.on] decorator, which turns a method in to a handler for the given message or event.
```
Comment (davep) [inline docs/guide/events.md:160]:
```suggestion
In addition to the naming convention, message handlers may be created with the [`on`][textual.on] decorator, which turns a method into a handler for the given message or event.
```

File: docs/guide/events.md [hunk 2]
```diff
@@ -155,6 +155,66 @@ Textual uses the following scheme to map messages classes on to a Python method.
...
+    3. Matches the button with an id of "quit"
+
+=== "Output"
+
+    ```{.textual path="docs/examples/events/on_decorator02.py"}
+    ```
+
+While there are a few more lines of code, it is clearer what will happen when you click any given button.
+
+Note that the decorator requires that the message class has a `control` attribute which should the widget associated with the message.
```
Comment (davep) [inline docs/guide/events.md:214]:
```suggestion
Note that the decorator requires that the message class has a `control` attribute which should be the widget associated with the message.
```

File: tests/test_on.py
Comment (davep) [inline tests/test_on.py:1]: I might be tempted to add a test here that `MyApp.my_custom_handler.__doc__` is what is expected, but that might be taking things too far.

**Review by davep — APPROVED** [review]
Couple of typo things and an extreme test suggestion but otherwise looking good! Looking forward to using this.

File: docs/examples/events/on_decorator.css [hunk 3]
```diff
@@ -0,0 +1,8 @@
+Screen {
+    align: center middle;
+    layout: horizontal;
+}
+
+Button {
+    margin:2  4;
```
Comment (rodrigogiraoserrao) [inline docs/examples/events/on_decorator.css:7]:
```suggestion
    margin: 2 4;
```

File: docs/guide/events.md [hunk 4]
```diff
@@ -155,6 +155,66 @@ Textual uses the following scheme to map messages classes on to a Python method.
...
+This is true of the builtin controls, but you may need to add `control` to any [custom messages](#custom-messages) you write.
```
Comment (rodrigogiraoserrao) [inline docs/guide/events.md:215]: The statement is "the **message** must have an attribute [...]", so we say it is true of the builtin _messages_:

```suggestion
This is true of the builtin messages, but you may need to add `control` to any [custom messages](#custom-messages) you write.
```

File: docs/guide/events.md
Comment (rodrigogiraoserrao) [inline docs/guide/events.md:1]: What if there is a generic handler `on_button_pressed` and a handler with the decorator `@on`?
What if multiple `@on` methods match?

File: src/textual/_on.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        ```python
+        @on(Button.Pressed, "#quit")
+        def quit_button(self) -> None:
+            self.app.quit()
+        ```
+
+    Args:
+        message_type: The message type (i.e. the class).
+        selector: An optional [selector](/guide/CSS#selectors). If supplied, the handler will only be called if `selector`
+            matches the sender of the message.
```
Comment (rodrigogiraoserrao) [inline src/textual/_on.py:31]: Might be a good idea to note that this sender is checked in the attribute `control`.

File: src/textual/_on.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+from .css.parse import parse_selectors
+from .css.tokenizer import TokenError
+from .message import Message
+
+DecoratedType = TypeVar("DecoratedType")
+
+
+class OnDecoratorError(Exception):
+    """The selector in the 'on' decorator failed to parse."""
```
Comment (rodrigogiraoserrao) [inline src/textual/_on.py:13]: This error is also used when other things fail, so maybe the docstring should be a bit more generic?

**Review by rodrigogiraoserrao — APPROVED** [review]
A couple of minor suggestions.

**Pull request merged**